src/index.js

```diff
@@ -252,56 +259,65 @@ function UseWalletProvider({
         })
 
       if (!web3ReactConnector) {
-        throw new UnsupportedConnectorError(connectorId)
+        setStatus('error')
+        setError(new UnsupportedConnectorError(connectorId))
+        return
       }
 
       try {
         // TODO: there is no way to prevent an activation to complete, but we
         // could reconnect to the last provider the user tried to connect to.
-        setActivating(connectorId)
+        setConnector(connectorId)
         await web3ReactContext.activate(web3ReactConnector, null, true)
-        setActivated(connectorId)
-        setActivating(null)
+        setStatus('connected')
       } catch (err) {
-        setActivating(null)
-
         // Don’t throw if another connection has happened in the meantime.
         if (id !== activationId.current) {
           return
         }
+
+        // If not, the error has been thrown during the current connection attempt,
+        // so it's correct to indicate that there has been an error
+        setConnector(null)
+        setStatus('error')
+
         if (err instanceof UnsupportedChainIdError) {
-          throw new UnsupportedChainError(-1, chainId)
+          setError(new UnsupportedChainError(-1, chainId))
+          return
         }
         // It might have thrown with an error known by the connector
         if (connector.handleActivationError) {
-          connector.handleActivationError(err)
+          const thrownError = connector.handleActivationError(err)
+          setError(thrownError)
+          return
         }
-        // Otherwise, throw the received error
-        throw err
+
+        // Otherwise, set to state the received error
+        setError(err)
       }
     },
-    [chainId, connectors, deactivate, web3ReactContext]
+    [chainId, connectors, reset, web3ReactContext]
   )
 
   useEffect(() => {
     if (!account || !ethereum) {
+      setStatus('disconnected')
       return
     }
 
     let cancel = false
 
-    setConnected(true)
     setIsContract(false)
 
     getAccountIsContract(ethereum, account).then(isContract => {
       if (!cancel) {
+        setStatus('connected')
```

Comment:
Did this moved here in order to get `status === 'connected'` at the same time than we get the proper `isContract`?

IIRC the idea was to connect as fast as possible, and update other things asynchronously (`balance` and `isContract`). But I agree that not knowing when the `isContract` information is available is a problem. What about having `isContract === null` by default, then setting it to `true` or `false` asynchronously?

Reply:
Yes, that's why; I was worried about this race condition ocurring for someone that may rely on this for their app. I definitely agree about the suggestion, but my worry is that people may not realize they may have a bug by doing something like:

```
// lots of code
if (!isContract) {
  // handle this case
}
```

When the account is _actually_ a contract account. What about having `isContract === null`, and having strings define if is a contract or not, like `'contract'` or `'normal'`? The naming's definitely not precise enough, but just an idea. :)

Reply:
Yes that sounds good! Should we have a name more like `type` or `accountType` then?

Reply:
Yep! `type` sounds good to me, as I think that the fact that it's coming from `use-wallet` is enough context.